### DISCUSSION TITLE
Help!  Results don't match TradingView!

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#801] [Help and support]
Many of our users will compare the results of our library to other charting platforms to both validate their implementation and the library itself.  It's a good idea!  There are quite a few libraries out there of varying quality and you want to make sure you're using a good one.  When results don't match, here are a few things to consider and a list of similar issues from other users that might help you find the reason (hint: it's not usually a bug in our library).

### Things to consider and validate

- #688 data is inaccurate.  Make sure you're using enough historical quotes to exceed warmup periods.
- **The underlying quote data may be different**.  If the underlying quote data is different in any way, the results will be different.  Different data providers may have slightly different prices; this is especially prevalent with cryptocurrencies that may aggregate with different cutoff times or may include or exclude weekends.
- **Use the exact same parameter arguments** as your basis for comparison.
- **Some indicators rely on the start date of history**, #433.  VWAP is another example.  You'd have to know and use the same start date for a proper comparison.
- **Your basis for comparison uses an altered formula**.  Keep in mind that some charting systems make creative modifications to formulas to make them work on their platforms.  Sometimes they have bugs!  We've found that TradingView is usually not a good source of truth for comparison, especially with community contributed indicators.  Compare your source with other third-party providers before assuming they're accurate.
- **Do the math**.  Use the raw quote data and do the math if you can.  We've manually calculated our formulas and do not copy other peoples' code in our implementation.  Our manual calculation spreadsheets can be found in our repo `tests` folder and are the basis for our unit tests.

### Related non-library issues and outcomes

- #318 - TradingView uses a non-standard formula (or has a bug)
- #22 - the user was using insufficient quote history
- #319, #431.  See also #433 - the user was not considering the starting date
- #654 - the other platform was using a non-standard formula
- #691 - the user was not handling data properly
- #607 - TradingView uses a non-standard formula (or has a bug)
- #384 - TradingView uses a non-standard formula (or has a bug)
- #353 - the user was not handling data properly
- #738 - the other platform had different quote data
- #820 - TradingView uses a non-standard formula (or has a bug)
- #1029 - TradingView uses a non-standard formula (or has a bug)
- #899 - the user was using different parameter values, and possibly different quotes to compare
- https://github.com/DaveSkender/Stock.Indicators/discussions/251#discussioncomment-5894003 - TradingView shows insufficient forward candles
- #1169 - the user was using insufficient quote history 

With that said, if you've confirmed that none of these situations are the reason for differences and you're able to narrow down to a problem in this library, please do [submit an issue](https://github.com/DaveSkender/Stock.Indicators/issues).  We always want to ensure our library is accurate and appreciate the opportunity to double-check.

**Comment by M22arius494** [comment]
I am comparing a lot lately with Tradingview, but for some reason Tradingview is sometimes even missing candles leading to mismatches. 

In this case it was a coin with low volume and there were frequently missing candles on the 5m chart.
